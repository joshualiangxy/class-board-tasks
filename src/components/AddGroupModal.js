import React, { useState } from 'react';
import Modal from 'react-modal';
import { startNewGroup } from '../actions/groups';
import { connect } from 'react-redux';

// TODO: Add the ability to add other users into the group
const AddGroupModal = ({ isOpen, onRequestClose, startNewGroup }) => {
  const [groupName, setGroupName] = useState('');
  const [module, setModule] = useState('');
  const [error, setError] = useState('');

  const onGroupNameChange = e => setGroupName(e.target.value);

  const onModuleChange = e => {
    const module = e.target.value.toUpperCase();

    setModule(module);
  };

  const onCancel = () => {
    setError('');
    setGroupName('');
    setModule('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedName = groupName.trim();
    const submittedMod = module.trim();
    if (!submittedName) {
      setError('Please enter a group name');
    } else {
      setError('');
      setGroupName('');
      setModule('');
      startNewGroup(submittedName, submittedMod).then(() => onRequestClose());
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={onCancel}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="Group name (required)"
          value={groupName}
          onChange={onGroupNameChange}
          autoFocus
        />
        <input
          type="text"
          placeholder="Module code/Class name (Optional)"
          value={module}
          onChange={onModuleChange}
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startNewGroup: (groupName, module) =>
    dispatch(startNewGroup(groupName, module))
});

export default connect(null, mapDispatchToProps)(AddGroupModal);
