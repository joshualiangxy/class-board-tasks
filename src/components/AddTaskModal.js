import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import { startAddPersonalTask, startAddGroupTask } from '../actions/tasks';
import TaskForm from './TaskForm';

export const AddTaskModal = ({
  startAddPersonalTask,
  startAddGroupTask,
  removeAdmin,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  addGroupTask,
  users
}) => {
  const submitTask = task =>
    gid
      ? startAddGroupTask(task, groupName)
          .then(isAdmin => {
            if (isAdmin) return addGroupTask(task);
            else return removeAdmin();
          })
          .then(() => onRequestClose())
      : startAddPersonalTask(task).then(() => onRequestClose());

  return (
    <Modal
      className="task-modal"
      isOpen={isOpen}
      contentLabel="Add Task"
      onRequestClose={onRequestClose}
      closeTimeoutMS={200}
      appElement={document.getElementById('root')}
    >
      <h1>Add Task</h1>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startAddPersonalTask: task => dispatch(startAddPersonalTask(task)),
  startAddGroupTask: (task, groupName) =>
    dispatch(startAddGroupTask(task, groupName))
});

export default connect(undefined, mapDispatchToProps)(AddTaskModal);
