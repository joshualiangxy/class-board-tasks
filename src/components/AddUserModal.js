import React, { useState } from 'react';
import Modal from 'react-modal';
import firebase from '../firebase/firebase';

// Note: This still works off uid for now. This also does NOT work yet.
const AddUserModal = ({ isOpen, onRequestClose, group }) => {
  const addNewUser = (userEmail, group) => {
    const uid = userEmail; // TODO: Change this when email-uid are linked!

    // Both of these references are DocumentReference
    const userRef = firebase.firestore().collection('users').doc(uid);
    const groupRef = firebase.firestore().collection('groups').doc(group);

    // Gets the data from a user.
    userRef.get().then(doc => {
      const data = doc.data();
      // Write displayName, studentNum, and UID into a new document
      if (typeof data === 'undefined') {
        // Most likely a user that doesn't exist.
        setError('No such user found');
      } else {
        groupRef
          .collection('users')
          .doc(uid) // Create document for this user's UID
          .set({
            studentNum: data.studentNum,
            displayName: data.displayName,
            uid: uid,
            admin: false
          })
          .then(() => {
            setError('User added!');
          })
          .catch(error => {
            setError('Something went wrong.');
          });
      }
      // Defaults to non-admin.
    });
  };

  // TODO: When calling this modal, pass in the group UID as a prop
  const [userEmail, setUserEmail] = useState('');
  const [error, setError] = useState('');
  // const [usersAdded, setUsersAdded] = useState([]); // For listing down people that have been added?

  const onUserEmailChange = e => {
    setUserEmail(e.target.value);
  };

  const onCancel = () => {
    setUserEmail('');
    onRequestClose();
  };

  const onSubmit = e => {
    e.preventDefault();
    const submittedEmail = userEmail.trim();
    if (!submittedEmail) {
      setError('Please enter a group name');
    } else {
      addNewUser(submittedEmail, group);
      // TODO: Catch user not found
      // Does not close after submitting, so that the user can continue to add more users

      // setUserEmail(''); // Also resets the field.
      // setError('Added!');
    }
  };

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="New Group"
      onRequestClose={() => onRequestClose()}
      appElement={document.getElementById('root')}
    >
      <form onSubmit={onSubmit}>
        <input
          type="text"
          placeholder="User ID (required)"
          value={userEmail}
          onChange={onUserEmailChange}
          autoFocus
        />
        <button>Submit</button>
        <button onClick={onCancel}>Cancel</button>
        <div>{error}</div>
      </form>
    </Modal>
  );
};

export default AddUserModal;
