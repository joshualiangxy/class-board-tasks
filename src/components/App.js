import React from 'react';
import '../styles/App.css';

const App = () => (
  <div>
    <h1>This is my app</h1>
  </div>
);

export default App;

