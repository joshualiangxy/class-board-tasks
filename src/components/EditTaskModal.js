import React from 'react';
import Modal from 'react-modal';
import { connect } from 'react-redux';
import {
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks
} from '../actions/tasks';
import TaskForm from './TaskForm';

export const EditTaskModal = ({
  startEditPersonalTask,
  startEditGroupTask,
  startSetTasks,
  isOpen,
  onRequestClose,
  gid,
  groupModule,
  groupName,
  task,
  users,
  editGroupTask
}) => {
  const submitTask = updates =>
    gid
      ? startEditGroupTask(task.id, updates, groupName).then(() => {
          editGroupTask(task.id, updates);
          startSetTasks(); // Temporary solution
          onRequestClose();
        })
      : startEditPersonalTask(task.id, updates).then(() => onRequestClose());

  return (
    <Modal
      isOpen={isOpen}
      contentLabel="Edit Task"
      onRequestClose={onRequestClose}
      appElement={document.getElementById('root')}
    >
      <h2>Edit Task</h2>
      <TaskForm
        gid={gid}
        groupModule={groupModule}
        users={users}
        submitTask={submitTask}
        onRequestClose={onRequestClose}
        task={task}
      />
    </Modal>
  );
};

const mapDispatchToProps = dispatch => ({
  startEditPersonalTask: (id, updates) =>
    dispatch(startEditPersonalTask(id, updates)),
  startEditGroupTask: (id, updates, groupName) =>
    dispatch(startEditGroupTask(id, updates, groupName)),
  startSetTasks: () => dispatch(startSetTasks())
});

export default connect(undefined, mapDispatchToProps)(EditTaskModal);
