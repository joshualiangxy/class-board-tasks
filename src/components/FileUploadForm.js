import React, { useState } from 'react';
import firebase from '../firebase/firebase';
import { startUpdateDownloadURL } from '../actions/tasks';
import { uploadFile, startRemoveUserFile } from '../actions/files';
import { connect } from 'react-redux';

export const FileUploadForm = ({
  id,
  uid,
  uploadFile,
  TaskEvent,
  startUpdateDownloadURL,
  startRemoveUserFile,
  fileExists
}) => {
  const [error, setError] = useState('');
  const [uploading, setUploading] = useState(false);

  const fileUpload = e => {
    e.preventDefault();
    e.persist();

    const element = document.getElementById(id);
    const files = element.files;
    if (files.length > 0) {
      const file = files[0];

      Promise.resolve()
        .then(() => {
          if (fileExists) return startRemoveUserFile(uid);
        })
        .then(() => {
          const uploadTask = uploadFile(file);

          uploadTask.on(
            TaskEvent.STATE_CHANGED,
            () => setUploading(true),
            error => {
              switch (error.code) {
                case 'storage/unauthorized':
                  setError('Not allowed to upload file!');
                  break;
                case 'storage/canceled':
                  setError('File upload cancelled');
                  break;
                default:
                  setError('File upload failed');
                  break;
              }
              setUploading(false);
            },
            () => {
              const ref = uploadTask.snapshot.ref;
              const fileName = ref.name;

              ref
                .getDownloadURL()
                .then(downloadURL => {
                  startUpdateDownloadURL(downloadURL, fileName);
                })
                .then(() => {
                  setError('');
                  setUploading(false);
                  e.target.reset();
                });
            }
          );
        });
    } else setError('No file selected');
  };

  return (
    <form onSubmit={fileUpload}>
      {error && <p>{error}</p>}
      <input
        type="file"
        id={id}
        onClick={e => e.stopPropagation()}
        accept=".pdf"
        disabled={uploading}
      />
      <button onClick={e => e.stopPropagation()} disabled={uploading}>
        {uploading ? 'Uploading...' : 'Upload'}
      </button>
    </form>
  );
};

const mapStateToProps = ({ tasks, auth }, { id }) => {
  const uid = auth.user.uid;
  const task = tasks.find(task => task.id === id);

  return {
    TaskEvent: firebase.storage.TaskEvent,
    fileExists: task ? task.downloadURLs.hasOwnProperty(uid) : false,
    uid
  };
};

const mapDispatchToProps = (dispatch, { id, gid, namingConvention }) => ({
  uploadFile: file => dispatch(uploadFile(file, id, namingConvention)),
  startUpdateDownloadURL: (downloadURL, fileName) =>
    dispatch(startUpdateDownloadURL(id, gid, downloadURL, fileName)),
  startRemoveUserFile: uid => dispatch(startRemoveUserFile(id, uid))
});

export default connect(mapStateToProps, mapDispatchToProps)(FileUploadForm);
