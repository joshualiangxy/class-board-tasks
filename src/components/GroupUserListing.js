import React from 'react';
import { connect } from 'react-redux';
import {
  kickUser,
  promoteUser,
  demoteUser,
  recheckAdmin
} from '../actions/groups';

/**
 * This component is used for GroupSettingsModal to display the
 * list of users, as well as the respective buttons for kicking
 * or editing.
 */
const GroupUserListing = ({
  users,
  setUsers,
  group,
  admin,
  setAdmin,
  uid,
  kickUser,
  kickUserLocal,
  setError
}) => {
  const onKick = (user, group) => {
    // Remove from users
    recheckAdmin(uid, group).then(isAdmin => {
      if (isAdmin) {
        kickUser(user, group)
          .then(() => kickUserLocal(user.uid))
          .then(() => setUsers(users.filter(u => u.uid !== user.uid)))
          .catch(error => console.log(error));
      } else {
        setAdmin(false);
        setError('You are not an admin!')
      }
    });
  };

  const onPromote = (user, group) => {
    recheckAdmin(uid, group).then(isAdmin => {
      if (isAdmin) {
        promoteUser(user, group).then(() => {
          setUsers(
            users.map(u => (u.uid === user.uid ? { ...u, admin: true } : u))
          );
        });
      } else {
        setAdmin(false);
        setError('You are not an admin!')
      }
    });
  };

  const onDemote = (user, group) => {
    recheckAdmin(uid, group).then(isAdmin => {
      if (isAdmin) {
        demoteUser(user, group).then(() => {
          setUsers(
            users.map(u => (u.uid === user.uid ? { ...u, admin: false } : u))
          );
        });
      } else {
        setAdmin(false);
        setError('You are not an admin!')
      }
    });
  };

  // The other buttons can only be seen if the current user is an admin.
  return users.map(user => (
    <div key={user.uid}>
      {user.displayName} {user.uid === uid && '(You)'}{' '}
      {admin && !user.admin && (
        <button onClick={() => onPromote(user, group)}>Make admin</button>
      )}{' '}
      {admin && user.admin && user.uid !== uid && (
        <button onClick={() => onDemote(user, group)}>Demote admin</button>
      )}
      {admin && user.uid !== uid && (
        <button onClick={() => onKick(user, group)}>Kick</button>
      )}
    </div>
  ));
};

const mapDispatchToProps = dispatch => ({
  kickUser: (user, group) => dispatch(kickUser(user, group))
});

export default connect(undefined, mapDispatchToProps)(GroupUserListing);
