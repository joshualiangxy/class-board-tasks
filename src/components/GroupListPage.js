import React, { useState } from 'react';
import AddGroupModal from './AddGroupModal';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';

export const GroupListPage = ({ groups }) => {
  const [open, setOpen] = useState(false);

  const onRequestClose = () => {
    setOpen(false);
  };

  const openNewGroup = () => {
    setOpen(true);
  };

  return (
    <div>
      <div className="page-header">
        <div className="content-container">
          <h1 className="page-header__title">Groups</h1>
          <button
            className="button button--norm page-header__actions"
            onClick={openNewGroup}
          >
            Add New Group
          </button>
        </div>
      </div>
      <div className="content-container">
        {
          // Add a 'no groups' when there's no groups
          groups.length === 0 ? (
            <p className="list-header__text">No groups</p>
          ) : (
            groups.map(group => (
              <div key={group.gid} className="grouplist">
                <Link
                  className="button--link-black"
                  to={`/groups/${group.gid}`}
                >
                  <h2 className="groupitem">{group.name}</h2>
                  
                </Link>
              </div>
            ))
          )
        }
      </div>

      <AddGroupModal isOpen={open} onRequestClose={onRequestClose} />
    </div>
  );
};

const mapStateToProps = state => ({ groups: state.groups });

export default connect(mapStateToProps)(GroupListPage);
