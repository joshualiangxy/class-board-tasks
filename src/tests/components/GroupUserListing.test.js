import React from 'react';
import { shallow } from 'enzyme';
import { GroupUserListing } from '../../components/GroupUserListing';
import groupUsers from '../fixtures/groupUsers';

const group = '72b6b3bb-cb74-4d08-ae67-dfc1b51baaf9';
const users = groupUsers;
const setUsers = jest.fn(users => true);
const admin = true;
const setAdmin = jest.fn();
const kickUserLocal = jest.fn();
const uid = 'testuid';
const kickUser = jest.fn((uid, gid) => Promise.resolve());
const setError = jest.fn(err => true);

let wrapper;

beforeEach(() => {
  jest.clearAllMocks();
  wrapper = shallow(
    <GroupUserListing
      users={users}
      setUsers={setUsers}
      group={group}
      admin={admin}
      setAdmin={setAdmin}
      kickUser={kickUser}
      kickUserLocal={kickUserLocal}
      setError={setError}
      uid={uid}
    />
  );
});

describe('render', () => {
  it('should render all the users in the group', () => {
    // 2 users: Robert and Peter
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find('.userwrapper')).toHaveLength(2);
  });

  it('should not render buttons for non-admins', () => {
    wrapper.setProps({ admin: false });
    expect(wrapper).toMatchSnapshot();
  });
});

describe('modifying users', () => {
  it('should promote a user', () => {
    expect(users[1].admin).toEqual(false);
    expect(wrapper).toMatchSnapshot();
    // console.log(wrapper.find('button').at(0).text())
    wrapper
      .find('button')
      .at(0)
      .prop('onClick')()
      .then(() => {
        const newusers = users.map(u =>
          u.uid === users[1].uid ? { ...u, admin: true } : u
        );
        expect(setUsers).toHaveBeenCalledTimes(1);
        expect(setUsers).toHaveBeenLastCalledWith(newusers);
        expect(newusers[1].admin).toEqual(true);
      });
  });

  it('should demote a user', () => {
    const users2 = [...groupUsers];
    users2[1].admin = true;
    wrapper.setProps({ users: users2 });
    expect(users2[1].admin).toEqual(true);
    expect(wrapper).toMatchSnapshot();
    // console.log(wrapper.find('button').at(0).text())
    wrapper
      .find('button')
      .at(0)
      .prop('onClick')()
      .then(() => {
        const newusers = users.map(u =>
          u.uid === users[1].uid ? { ...u, admin: false } : u
        );
        expect(setUsers).toHaveBeenCalledTimes(1);
        expect(setUsers).toHaveBeenLastCalledWith(newusers);
        expect(newusers[1].admin).toEqual(false);
      });
  });
  it('should kick users', () => {
    const result = users.filter(u => u.uid === 'testuid');
    expect(wrapper.find('.userwrapper')).toHaveLength(2);
    wrapper
      .find('button')
      .at(1)
      .prop('onClick')()
      .then(() => {
        expect(kickUser).toHaveBeenLastCalledWith(users[1], group);
        expect(kickUserLocal).toHaveBeenLastCalledWith('testuid2');
        expect(setUsers).toHaveBeenLastCalledWith(result);
      });
  });
});
