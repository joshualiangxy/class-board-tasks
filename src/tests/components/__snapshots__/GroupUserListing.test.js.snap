// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modifying users should demote a user 1`] = `
<div
  className="bordertop"
>
  <div
    className="userwrapper"
    key="testuid"
  >
    <div>
      <p
        className="list-header__text"
      >
        Robert
        (You)
      </p>
    </div>
    <div />
  </div>
  <div
    className="userwrapper"
    key="testuid2"
  >
    <div>
      <p
        className="list-header__text"
      >
        Peter
      </p>
    </div>
    <div>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Demote admin
      </button>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Kick
      </button>
    </div>
  </div>
</div>
`;

exports[`modifying users should promote a user 1`] = `
<div
  className="bordertop"
>
  <div
    className="userwrapper"
    key="testuid"
  >
    <div>
      <p
        className="list-header__text"
      >
        Robert
        (You)
      </p>
    </div>
    <div />
  </div>
  <div
    className="userwrapper"
    key="testuid2"
  >
    <div>
      <p
        className="list-header__text"
      >
        Peter
      </p>
    </div>
    <div>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Make admin
      </button>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Kick
      </button>
    </div>
  </div>
</div>
`;

exports[`render should not render buttons for non-admins 1`] = `
<div
  className="bordertop"
>
  <div
    className="userwrapper"
    key="testuid"
  >
    <div>
      <p
        className="list-header__text"
      >
        Robert
        (You)
      </p>
    </div>
    <div />
  </div>
  <div
    className="userwrapper"
    key="testuid2"
  >
    <div>
      <p
        className="list-header__text"
      >
        Peter
      </p>
    </div>
    <div />
  </div>
</div>
`;

exports[`render should render all the users in the group 1`] = `
<div
  className="bordertop"
>
  <div
    className="userwrapper"
    key="testuid"
  >
    <div>
      <p
        className="list-header__text"
      >
        Robert
        (You)
      </p>
    </div>
    <div />
  </div>
  <div
    className="userwrapper"
    key="testuid2"
  >
    <div>
      <p
        className="list-header__text"
      >
        Peter
      </p>
    </div>
    <div>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Make admin
      </button>
      <button
        className="button button--norm-small button--right"
        onClick={[Function]}
      >
        Kick
      </button>
    </div>
  </div>
</div>
`;
